### ISSUE TITLE
how to upload file ? (sheep_counter example)

### ISSUE THREAD

**Issue by akhileshsarda** [babelouest/ulfius#98]
Hello,
In a sheep_counter example, I find there is a facility to upload file from client side (browser)
and we can receive it at server side. (file_upload_callback).
So, I wanted to save that file and I tryied some file handling operations, but
I faced one problem that whenever I am uploading file containing 987 and more chars 
it's misbehaving (some random data gets added). (for less than 986, working fine).
Could you please help me in this ?

**Comment by babelouest** [comment]
Hello @akhileshsarda ,

I'm having some difficulties to reproduce your bug, but concerning file uploads here are some precisions.

By default, a file upload is considered as a POST parameter like any other POST parameters.
If you upload binary files, you may have difficulties because by default, ulfius ignores parameter values that are not `utf8`. You can disable this by setting `struct _u_instance.check_utf8` to `0`.

If you use `ulfius_set_upload_file_callback_function`, you can specify a file upload callback function that will be called prior to any other callback function. This function will be called as many time as possible until all the file data is parsed by the framework. When using `ulfius_set_upload_file_callback_function`, the file data will not be available in the `struct _u_request.map_post_body` structure and you won't need to set `struct _u_instance.check_utf8` to `0`.

Conecrning your problem with files bigger than 986 bytes, I have no Idea, my tests work fine with larger files. If you still have the problem, can you post a sample code file to reproduce the problem and a file you're uploading but has problems?

**Comment by akhileshsarda** [comment]
See, what i have added in the code:
```
/**
 * File upload callback function
 */
int file_upload_callback (const struct _u_request * request,
                          const char * key,
                          const char * filename,
                          const char * content_type,
                          const char * transfer_encoding,
                          const char * data,
                          uint64_t off,
                          size_t size,
                          void * cls) {
  y_log_message(Y_LOG_LEVEL_DEBUG, "Got from file '%s' of the key '%s', offset %llu, size %zu, strlen(data) %u, cls is '%s'", filename, key, off, size,strlen(data), cls);
  FILE * file;

  file=fopen("download.md","a");
  fputs(data,file);
  fclose(file);
  return U_OK;
}
```
and sending README.md from client side.
Here's what i get on server side:
![sheep](https://user-images.githubusercontent.com/34626659/52167297-fd249700-273e-11e9-86f1-62519b728034.png)


There is a difference in size and string length.
In received file I observed, repeated characters at some point as shown in attached pic.
see the diff of both files:
![diff](https://user-images.githubusercontent.com/34626659/52167318-63111e80-273f-11e9-9d6a-ca92cfdc9cc6.png)
The "your" part has been repeated.

**Comment by babelouest** [comment]
I think I understand what's happening.

You use `fputs(data,file);` to write into the file but you assume `data` is exactly `size` lenght, where in fact, for some reason, libmicrohttpd underneath puts more characters at the end of `data`.

if I replace the `fputs` statement with `fwrite(data, size, 1, file);`, the file `download.md` is correct.

Can you confirm my hypothesis?

**Comment by akhileshsarda** [comment]
Yes, that's correct!
I was using fputs(data,file) which was same as fwrite(data,strlen(data),1,file);
But data wasn't equal to size, as few more chars were added to data. Now it's working fine.

One more question I had, this callback invokes after every 1024 bytes of data transfer (i guess).
Can we increase this limit ? (so that 2048 bytes or more can be sent at a time)
If yes then how ?

*Actually I wanted to add one feature to auth_example, as in case if required we could send client-server certificate or keys (depending upon the implementation & requirement) to peer side.
This might be helpful for validity (or some other) extension of certificates.

**Comment by akhileshsarda** [comment]
Is this correct way to send/upload files (.crt .key) to a server (using file handling operations)?
Or else any better way?

**Comment by babelouest** [comment]
I'm not sure I understand your feature request but it's **never** a good idea to send private keys via a network.

But if you have a new feature request, I suggest you open a new issue and we discuss it there, therefore the conversations will be less complicated.

Also, if you agree that this issue about file upload is fixed, we can close it now.

**Comment by akhileshsarda** [comment]
Okay, if required I'll open a new issue.

What about the callback which invokes after every 1024 bytes of data transfer?
Can we increase this limit?
And we can close this issue.

**Comment by babelouest** [comment]
You can change the value of your `struct _u_response.stream_block_size` to increase this.
https://github.com/babelouest/ulfius/blob/master/API.md#response-structure

It can't be less than 256 though.

**Comment by babelouest** [comment]
Yep, I realized afterwards that `struct _u_response.stream_block_size` isn't used in upload file callbacks.
In your case, you must modify `ULFIUS_POSTBUFFERSIZE`: https://github.com/babelouest/ulfius/blob/master/include/ulfius.h#L1286

Should work better

**Comment by akhileshsarda** [comment]
Don't know i'm making mistake or not..
But both of them are not working..
Did you try it?

**Comment by akhileshsarda** [comment]
Where did you make that change that value? I changed it in ulfius.h :

> ./include/ulfius.h:1286: #define ULFIUS_POSTBUFFERSIZE 2048

As above not worked..

**Comment by babelouest** [comment]
That's how I made the change.

Are you sure you recompiled the source code and use the new compiled library?

**Comment by akhileshsarda** [comment]
Ohh! I forgot to do #make install 
Thanks for it.
Now working fine.
